### PR TITLE
Notification On Job Start

### PR DESCRIPTION
**Pull request by beeankha**
##### SUMMARY

Add feature for notifications to trigger on job start.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - API
 - UI

##### ADDITIONAL INFORMATION

Related to issue https://github.com/ansible/awx/issues/79

### CODE REVIEW

File: awx/api/views/__init__.py [hunk 1]
```diff
@@ -2143,6 +2152,11 @@ def post(self, request, *args, **kwargs):
         return super(InventorySourceNotificationTemplatesAnyList, self).post(request, *args, **kwargs)
 
 
+class InventorySourceNotificationTemplatesStartedList(InventorySourceNotificationTemplatesAnyList):
```
Comment (ryanpetrello) [inline awx/api/views/__init__.py:2114]: So it's kind of interesting to me that the views defined *here* for inventory source subclass the `AnyList` implementation, and **only** override the `relationship` value, but (see related comment above)...

File: awx/api/views/__init__.py [hunk 2]
```diff
@@ -752,6 +753,14 @@ class ProjectNotificationTemplatesAnyList(SubListCreateAttachDetachAPIView):
     relationship = 'notification_templates_any'
 
 
+class ProjectNotificationTemplatesStartedList(SubListCreateAttachDetachAPIView):
+
+    model = models.NotificationTemplate
```
Comment (ryanpetrello) [inline awx/api/views/__init__.py]: (continued from [comment below](https://github.com/ansible/awx/pull/3982/files#r289527169)) ...this one doesn't.  Maybe we should take this opportunity to update **all** of these views to follow the pattern where they inherit from the `AnyList` view?  Seems cleaner/less code to me.

Seems like we could update this entire file to follow this pattern (make everything, even the existing `SuccessList` and `ErrorList` implementations follow this pattern).

File: awx/main/models/notifications.py [hunk 3]
```diff
@@ -233,6 +236,8 @@ def send_notification_templates(self, status_str):
         if notification_templates:
             if status_str == 'succeeded':
                 notification_template_type = 'success'
+            elif status_str == 'running':
+                notification_template_type = 'started'
             else:
                 notification_template_type = 'error'
             all_notification_templates = set(notification_templates.get(notification_template_type, []) + notification_templates.get('any', []))
```
Comment (ryanpetrello) [inline awx/main/models/notifications.py]: I could see an argument for *not* including the `_any` relationship when `status_str="running"`.  If you think about what "any" means today, it really means "at the end, regardless of success or failure".

If you kept this code as-is, and had an "any" template defined, it would run *twice*.

1.  When the job went into `running` state, it would run.
2. When the job finished and went into either `success` or `failed` it would run.

Maybe we should change this logic to something like:

```python
    if status_str == 'succeeded':
        notification_template_type = 'success'
    elif status_str == 'running':
        notification_template_type = 'started'
    else:
        notification_template_type = 'error'
    all_notification_templates = set(notification_templates.get(notification_template_type, [])
    if status_str != 'running':
        all_notification_templates.update(notification_templates.get('any', [])))
```

File: awx/main/scheduler/task_manager.py [hunk 4]
```diff
@@ -587,4 +587,4 @@ def schedule(self):
 
                 # Operations whose queries rely on modifications made during the atomic scheduling session
                 for wfj in WorkflowJob.objects.filter(id__in=finished_wfjs):
-                    wfj.send_notification_templates('succeeded' if wfj.status == 'successful' else 'failed')
+                    wfj.send_notification_templates('started' if wfj.status == 'running' else ('succeeded' if wfj.status == 'successful' else 'failed'))
```
Comment (ryanpetrello) [inline awx/main/scheduler/task_manager.py]: I don't think this is where this belongs.  I think this code is explicitly handling workflow jobs that *are finished*?  So this is probably _not_ where we want to send "on start" notifications

@chrismeyersfsu what do you think is a reasonable place in the task manager to fire off "on start" notifications for workflow jobs?  What about here?

https://github.com/ansible/awx/blob/b1d75327e3b6bb146da8761f285f6b45608505ad/awx/main/scheduler/task_manager.py#L235

@chrismeyersfsu what do you think this feature actually means in the context of a WFJT?

File: awx/main/tasks.py [hunk 5]
```diff
@@ -314,7 +314,7 @@ def send_notifications(notification_list, job_id=None):
         update_fields = ['status', 'notifications_sent']
         try:
             sent = notification.notification_template.send(notification.subject, notification.body)
-            notification.status = "successful"
+            notification.status = "successful" or "running"
```
Comment (ryanpetrello) [inline awx/main/tasks.py]: Notifications have a `status`, and it represents whether or not sending the notification "worked".  They only have a few states:

image

I think this change is unnecessary.

File: awx/main/tasks.py [hunk 6]
```diff
@@ -1102,6 +1102,8 @@ def run(self, pk, **kwargs):
         self.instance = self.update_model(pk, status='running',
                                           start_args='')  # blank field to remove encrypted passwords
 
+        self.instance.send_notification_templates("running")
```
Comment (ryanpetrello) [inline awx/main/tasks.py]: Yep, I think this is a reasonable place to do this.  @AlanCoding @chrismeyersfsu you have any opinions?

File: awx/main/models/notifications.py (hunk 3 above)
Comment (wenottingham) [reply awx/main/models/notifications.py]: Note that 'any' is currently an API-only construct that's not in the UI. TBD how much it's used in practice.

Is it worth migrating 'any' to 'finished'?

Comment (ryanpetrello) [reply awx/main/models/notifications.py]: Possibly, but I'd advocate for it in another PR (it'd maybe involve an additional migration to rename the tables if we wanted to be consistent).  The idea of "any" is a little confusing now, but at a minimum, let's avoid any behavioral changes as a result of this new functionality.

I'm kind of back and forth on this, because renaming it would represent a breaking API change.  I'm also betting not many people know about or use this.

**Review by ryanpetrello — COMMENTED** [review]
~It would _really_ be ideal to add some API tests somewhere in `awx/main/tests/functional/api/` that verified (for JobTemplates, Inventory, SystemJobTemplates, WorkflowJobTemplates, etc...):~

~1.  You can do an `HTTP GET` to the appropriate new endpoints (e.g., `/api/v2/<resource>/notification_templates_started/`) and get an HTTP 200 with zero results.
2.  That you can *attach* a notification template with an `HTTP POST` and `{"id": X, "associate": true}`, and then _see_ the results with `HTTP GET` at the `notification_templates_started` endpoint.~

**^ DONE**

Beyond that, I'm of the opinion that testing that the notifications are properly **created** at the right time is a better fit for integration tests (i.e., if I configure a template for "on start" for a JT/Inventory/WF/whatever, when <that thing> actually *enters running state*, I see a notification get created at `/api/v2/notifications/`).  We'll want to explicitly test this for *actual unified jobs that run* (e.g., playbooks, inventory, etc...) and _also_ for when **an actual workflow starts**.  @one-t @Spredzy what do _you_ think about this?

Another thing I think we should *explicitly* test is that this all behaves correctly according to _this_ comment:  https://github.com/ansible/awx/pull/3982/files#r289535364

File: awx/main/tasks.py (hunk 6 above)
Comment (AlanCoding) [reply awx/main/tasks.py]: Had to look at the code for `send_notification_templates`, but it calls `.delay` on the relevant task to _actually_ send the notification templates. So that is what we want, because it makes it congruent with `websocket_emit_status`, where we create some _internal_ network traffic within the cluster. That looks good.

I would slightly prefer that both of these methods move just a few lines down to inside of the `try:` block. For anything that isn't trivial variable initialization, I would prefer for errors to be surfaced in the job traceback.

File: awx/main/tasks.py (hunk 5 above)
Comment (ghjm) [reply awx/main/tasks.py]: > notification.status = "successful" or "running"

Am I missing something here, or isn't this just always going to assign `"successful"` anyway?  Since `s1 or s2` will only evaluate to `s2` in the case where `s1` is empty, and `"successful"` isn't empty.

Comment (ryanpetrello) [reply awx/main/tasks.py]: @ghjm 

Nope.  Syntactically, this will *always* evaluate to `notification.status = "successful"`.

This code should just be removed.  I think it was just copypasta on @beeankha's part.

File: awx/main/tasks.py (hunk 6 above)
Comment (ryanpetrello) [reply awx/main/tasks.py]: @beeankha I agree with @AlanCoding.  Let's move this function call lower into the `try:` block below.  Maybe right before the `self.pre_run_hook()` call?

File: awx/main/tasks.py (hunk 5 above)
Comment (beeankha) [reply awx/main/tasks.py]: This was a case of `"running" status` = 🔨 and any place with that logic (for notifications) looking like a nail.  My strategy was to overdo it a bit and then prune back redundancy as needed.  Will remove this code!

File: awx/main/scheduler/task_manager.py (hunk 4 above)
Comment (AlanCoding) [reply awx/main/scheduler/task_manager.py]: If we add the on-start notifications to that line, then we should move this _existing code_ for the failed / success case up to around:

https://github.com/ansible/awx/blob/b1d75327e3b6bb146da8761f285f6b45608505ad/awx/main/scheduler/task_manager.py#L195

I've sat next to them while @chrismeyersfsu and @beeankha were troubleshooting this. I won't spoil the ending for you, but I'm pretty sure the trouble they are having is the same reason that _this_ code was put in this place instead of the more logical location.

File: awx/api/views/__init__.py (hunk 2 above)
Comment (beeankha) [reply awx/api/views/__init__.py]: I went ahead and updated this file with that pattern!

File: awx/main/tasks.py (hunk 6 above)
Comment (beeankha) [reply awx/main/tasks.py]: Done!

File: awx/main/scheduler/task_manager.py [hunk 7]
```diff
@@ -233,6 +233,7 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
         else:
             if type(task) is WorkflowJob:
                 task.status = 'running'
+                task.send_notification_templates('running') # <----
```
Comment (beeankha) [inline awx/main/scheduler/task_manager.py]: Oops, I left a bookmark comment in here, will remove.

File: awx/api/views/organization.py [hunk 8]
```diff
@@ -181,6 +181,14 @@ class OrganizationNotificationTemplatesAnyList(SubListCreateAttachDetachAPIView)
...
 
 
+class OrganizationNotificationTemplatesStartedList(SubListCreateAttachDetachAPIView):
+
+    model = NotificationTemplate
+    serializer_class = NotificationTemplateSerializer
+    parent_model = Organization
+    relationship = 'notification_templates_started'
+
+
```
Comment (ryanpetrello) [inline awx/api/views/organization.py:187]: If we're gonna move to new pattern, let's do the same thing with `OrganizationNotificationTemplatesAnyList`, `OrganizationNotificationTemplatesStartedList`, `OrganizationNotificationTemplatesErrorList `

Comment (beeankha) [reply awx/api/views/organization.py:187]: Oops, missed these.  They've been changed.  I searched the codebase for other instances and it seems like this was the last bit.

File: awx/main/scheduler/task_manager.py [hunk 9]
```diff
@@ -233,6 +235,7 @@ def start_task(self, task, rampart_group, dependent_tasks=None, instance=None):
         else:
             if type(task) is WorkflowJob:
                 task.status = 'running'
+                task.send_notification_templates('running')
```
Comment (AlanCoding) [inline awx/main/scheduler/task_manager.py:238]: Don't you need the same kind of `connection.on_commit(lambda: workflow_job.send_notification_templates('running'))` here as well?

Also, I can't understand why we have an exception of "succeeded" (as opposed to "successful") for a finished job, but pass in "running" for the on-start case, since the user-facing terminology is "started"

File: awx/main/models/notifications.py [hunk 10]
```diff
@@ -233,14 +237,20 @@ def send_notification_templates(self, status_str):
...
                 notification_template_type = 'error'
-            all_notification_templates = set(notification_templates.get(notification_template_type, []) + notification_templates.get('any', []))
-            if len(all_notification_templates):
-                try:
-                    (notification_subject, notification_body) = getattr(self, 'build_notification_%s_message' % status_str)()
-                except AttributeError:
-                    raise NotImplementedError("build_notification_%s_message() does not exist" % status_str)
+            all_notification_templates = set(notification_templates.get(notification_template_type, []))
+            if status_str != 'running':
+                all_notification_templates.update(notification_templates.get('any', []))
```
Comment (AlanCoding) [inline awx/main/models/notifications.py]: It looks like this behavior should go in the docs. If you attach a notification template as an "any" NT, that means that it'll send a notification as it completes. So it overlaps with success & failure, but not with start.

I see how that's sensible, but it's not obvious from the naming, and the docs should specifically enumerate when each type of notification template will be sent.

File: awx/main/models/notifications.py [hunk 11]
```diff
@@ -233,14 +237,20 @@ def send_notification_templates(self, status_str):
...
+            try:
+                (notification_subject, notification_body) = getattr(self, 'build_notification_%s_message' % status_str)()
+            except AttributeError:
+                raise NotImplementedError("build_notification_%s_message() does not exist" % status_str)
+
+            def send_it():
                 send_notifications.delay([n.generate_notification(notification_subject, notification_body).id
                                           for n in all_notification_templates],
                                          job_id=self.id)
+            connection.on_commit(send_it)
```
Comment (AlanCoding) [inline awx/main/models/notifications.py]: this is kind of weird to add the on_commit here as well as in those other places. I'd prefer to just just it once.

File: awx/main/models/notifications.py [hunk 12]
```diff
@@ -233,14 +237,20 @@ def send_notification_templates(self, status_str):
         if notification_templates:
             if status_str == 'succeeded':
                 notification_template_type = 'success'
+            elif status_str == 'running':
+                notification_template_type = 'started'
             else:
                 notification_template_type = 'error'
-            all_notification_templates = set(notification_templates.get(notification_template_type, []) + notification_templates.get('any', []))
-            if len(all_notification_templates):
```
Comment (AlanCoding) [inline awx/main/models/notifications.py]: why did you remove `if len(all_notification_templates):`, the check for the no-op case? I certainly wouldn't want methods getting added to the on_commit list that don't do anything.

File: awx/main/models/notifications.py (hunk 11 above)
Comment (ryanpetrello) [reply awx/main/models/notifications.py]: Agreed.  @beeankha @chrismeyersfsu @AlanCoding how do you feel about *just* using the `on_commit` _here_ and get ridding of it in the other spots?  After all, the point is to no send the AMQP message until _after_ the web request transaction is over.

File: awx/main/models/notifications.py (hunk 10 above)
Comment (ryanpetrello) [reply awx/main/models/notifications.py]: Agreed.  I think this behavior makes the *most* sense, but it's also a little unclear that "all" really is more like "when the job finishes" cc @tvo318

Comment (ryanpetrello) [reply awx/main/models/notifications.py]: I'll also say, I sort of dislike this "any" feature because it's basically undocumented, doesn't exist in the UI, and you can accomplish it by just using "success" *and* "failure" together.  @matburt @wenottingham what if we just got rid of it?

File: awx/main/models/notifications.py (hunk 11 above)
Comment (AlanCoding) [reply awx/main/models/notifications.py]: That's fine with me, I have no opinion about which place it is put.

> until after the web request transaction is over.

 transactions defined manually within task-based actions, this is never called from a web request. But agree otherwise. As @chrismeyersfsu often points out, if you're not in a transaction (job run) then it runs right away, which is correct.

File: awx/main/models/notifications.py (hunk 10 above)
Comment (wenottingham) [reply awx/main/models/notifications.py]: You'd need to migrate existing ones. But not against it.

File: awx/main/models/notifications.py (hunk 12 above)
Comment (beeankha) [reply awx/main/models/notifications.py]: Oops! I 🔪 too much!!

File: awx/main/models/notifications.py (hunk 11 above)
Comment (beeankha) [reply awx/main/models/notifications.py]: Redundancy has been removed!

File: awx/main/models/notifications.py (hunk 10 above)
Comment (AlanCoding) [reply awx/main/models/notifications.py]: I'm also not against it, but if it happens, we should handle it as a separate pull request. For now, I think the main thing this needs before merging is content in the `docs/` folder & verification.

**Review by jbradberry — APPROVED** [review]
Nothing jumps out at me here.  :rocket:

File: awx/main/tests/functional/api/test_notifications.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+import pytest
```
Comment (beeankha) [inline awx/main/tests/functional/api/test_notifications.py:1]: Please note that I will be adding more tests to this file, it's just a placeholder for now (adding things piecemeal as they pass)!

**Review by AlanCoding — APPROVED** [review]

**Review by ryanpetrello — APPROVED** [review]
This looks great.  Once we add some integration tests like we talked about, I think this is ready to merge cc @Spredzy https://github.com/ansible/awx/pull/3982#pullrequestreview-244501462

File: awx/api/urls/system_job_template.py [hunk 14]
```diff
@@ -24,7 +24,7 @@
     url(r'^(?P<pk>[0-9]+)/schedules/$', SystemJobTemplateSchedulesList.as_view(), name='system_job_template_schedules_list'),
     url(r'^(?P<pk>[0-9]+)/notification_templates_any/$', SystemJobTemplateNotificationTemplatesAnyList.as_view(),
         name='system_job_template_notification_templates_any_list'),
-    url(r'^(?P<pk>[0-9]+)/notification_templates_any/$', SystemJobTemplateNotificationTemplatesStartedList.as_view(),
+    url(r'^(?P<pk>[0-9]+)/notification_templates_started/$', SystemJobTemplateNotificationTemplatesStartedList.as_view(),
```
Comment (ryanpetrello) [inline awx/api/urls/system_job_template.py:24]: Ooh, good eye on the typo.

File: awx/main/models/notifications.py (hunk 10 above)
Comment (beeankha) [reply awx/main/models/notifications.py]: New related AWX issue has been created here: https://github.com/ansible/awx/issues/4022

**Comment by beeankha** [comment]
I have tested the migration file by setting notification instances in the `any` state and then applying the migration.  Notifications in the `any` state are getting assigned to `success` and `error` (as intended) and are showing up as such in the API as well as the UI.

SuccessError

**Review by Spredzy — APPROVED** [review]

File: awx/main/migrations/0081_v360_notify_on_start.py [hunk 15]
```diff
@@ -22,7 +37,8 @@ class Migration(migrations.Migration):
             name='notification_templates_started',
             field=models.ManyToManyField(blank=True, related_name='unifiedjobtemplate_notification_templates_for_started', to='main.NotificationTemplate'),
         ),
-#        migrations.RunPython() stuff goes in here
+        # Separate out "any" notifications into "success" and "error" before the "any" state gets deleted.
+        migrations.RunPython(forwards_split_unified_job_template_any, forwards_split_organization_any),
```
Comment (ryanpetrello) [inline awx/main/migrations/0081_v360_notify_on_start.py:41]: This is wrong, you want two *separate* operations.  The second argument is the code to run if you *reversed* the migration (to roll back to a `0080`.

https://docs.djangoproject.com/en/2.2/ref/migration-operations/#runpython

You want:

```python
migrations.RunPython(forwards_split_unified_job_template_any, None),
migrations.RunPython(forwards_split_organization_any, None),
```

File: awx/main/migrations/0081_v360_notify_on_start.py [hunk 16]
```diff
@@ -5,6 +5,21 @@
 from django.db import migrations, models
 
 
+def forwards_split_unified_job_template_any(apps, schema_editor):
+    UnifiedJobTemplate = apps.get_model('main', 'unifiedjobtemplate')
+    for ujt in UnifiedJobTemplate.objects.all():
```
Comment (ryanpetrello) [inline awx/main/migrations/0081_v360_notify_on_start.py]: 🎉

Comment (ryanpetrello) [inline awx/main/migrations/0081_v360_notify_on_start.py:10]: 🎉

File: awx/main/models/notifications.py [hunk 17]
```diff
@@ -243,8 +243,6 @@ def send_notification_templates(self, status_str):
                 notification_template_type = 'error'
             all_notification_templates = set(notification_templates.get(notification_template_type, []))
             if len(all_notification_templates):
-                # if status_str != 'running':
```
Comment (ryanpetrello) [inline awx/main/models/notifications.py]: Let's go ahead and just delete this commented code.  If we ever need it for some reason, we can always bring it back with source control :)

File: awx/main/tests/functional/api/test_inventory.py [hunk 18]
```diff
@@ -42,11 +42,6 @@ def test_inventory_source_notification_on_cloud_only(get, post, inventory_source
 
     not_is = inventory_source_factory("not_ec2")
 
-    # Delete the below portion? ******
```
Comment (ryanpetrello) [inline awx/main/tests/functional/api/test_inventory.py]: Let's delete this.

File: awx/main/tests/functional/test_notifications.py [hunk 19]
```diff
@@ -92,29 +92,7 @@ def test_inherited_notification_templates(get, post, user, organization, project
     isrc.save()
     jt = JobTemplate.objects.create(name='test', inventory=i, project=project, playbook='debug.yml')
     jt.save()
-    # What can we replace the below tests with? ******
```
Comment (ryanpetrello) [inline awx/main/tests/functional/test_notifications.py]: ...and these.

File: awx/main/tests/functional/test_notifications.py [hunk 20]
```diff
@@ -107,12 +107,13 @@ def test_inherited_notification_templates(get, post, user, organization, project
     # assert len(isrc.notification_templates['any']) == 1
 
 
-@pytest.mark.django_db
-def test_notification_template_merging(get, post, user, organization, project, notification_template):
-    user('admin-poster', True)
-    organization.notification_templates_any.add(notification_template)
-    project.notification_templates_any.add(notification_template)
-    assert len(project.notification_templates['any']) == 1
+# Delete the test below? ******
```
Comment (ryanpetrello) [inline awx/main/tests/functional/test_notifications.py]: Let's just delete this.

File: awx/main/migrations/0081_v360_notify_on_start.py (hunk 15 above)
Comment (beeankha) [reply awx/main/migrations/0081_v360_notify_on_start.py:41]: This is getting fixed here: https://github.com/ansible/awx/pull/4098
